### PR TITLE
feature: Added support for sensors to be discovered from yaml

### PR DESCRIPTION
**Pull request by laf**
DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`

Just pushing, nothing more right now.

### CODE REVIEW

**Comment by laf** [comment]
Ok this is open for feedback now.

@murrant I've removed the oid_name section in the last commit that you don't like :P

**Comment by murrant** [comment]
Looks amazing laf :)

**Comment by FTBZ** [comment]
YAML type is working only for OIDs in table or can be used for all kinds?

**Comment by laf** [comment]
How do you mean @FTBZ?

**Comment by laf** [comment]
Do you mean can it be used for OIDs that aren't in a table? No not at the moment. If you have an example device that this doesn't work on then I can look at seeing what we can do.

**Comment by FTBZ** [comment]
Exactly, I have already had cases where the OIDs of sensors are not in tables but I do not remember which. Good to know for further integration.

**Comment by murrant** [comment]
If you snmpwalk a single item, doesn't basically return a single getnext?

File: includes/discovery/sensors.inc.php [hunk 1]
```diff
@@ -12,6 +12,17 @@
     d_echo($pre_cache);
 }
 
+if (isset($device['dynamic_discovery']['modules']['sensors'])) {
+    foreach ($device['dynamic_discovery']['modules']['sensors'] as $key => $data_array) {
+        foreach ($data_array as $data) {
+            foreach ((array)$data['oid'] as $oid) {
+                $tmp_name = $data['oid'];
+                $pre_cache[$tmp_name] = snmpwalk_cache_oid($device, $oid, $pre_cache[$tmp_name], $device['dynamic_discovery']['mib'], null, '-OeQUs');
```
Comment (murrant) [inline includes/discovery/sensors.inc.php]: You should check if the oid has already been fetched here.

Comment (laf) [inline includes/discovery/sensors.inc.php]: Done.

Walking a single entry is ok but I'm not sure how well the code will work against that yet. I've added support for passing a description as a string as well as saying which key in the array contains the description so it's possibly ok with single OIDs.

**Comment by laf** [comment]
Rebased. Should pass travis now.

**Comment by maciejkaczkowski** [comment]
1. Is possible to add {{ $index }} to descr? I have device without descr of ports:
.1.3.6.1.4.1.11195.2.1.10.2.1.12.1 = INTEGER: -68
.1.3.6.1.4.1.11195.2.1.10.2.1.12.2 = INTEGER: -57
.1.3.6.1.4.1.11195.2.1.11.2.1.27.1 = INTEGER: 100
.1.3.6.1.4.1.11195.2.1.11.2.1.27.2 = INTEGER: 90

Tuner Input Level 1 = -68
Tuner Input Level 2 = -57
Modulator Output Level 1 = 100
Modulator Output Level 2 = 90

2. Is possible to add option to skip if empty or skip if value = xxx?

**Comment by laf** [comment]
I've added support for using index in descr.

Can you give me an example on what you'd want to skip?

**Comment by maciejkaczkowski** [comment]
Thank you. 
For example Arris C4, if card is not inserted to chassis then value is 999
https://github.com/librenms/librenms/blob/master/includes/discovery/sensors/temperature/cmts.inc.php

**Comment by laf** [comment]
@maciejkaczkowski I've added that support:

skip_values:

**Comment by laf** [comment]
This should be good for a merge now imho.

**Comment by murrant** [comment]
Looks good to me laf, I plan to try to convert one of the devices I have, but we don't have to wait on me to merge.

**Comment by Rosiak** [comment]
Almost there!
However it only adds one of the multiple sensors under otherStateSensorTable:
Working
screen shot 2017-06-26 at 20 00 14
This PR:
screen shot 2017-06-26 at 20 00 20

Output from SNMP from this table:
```
otherStateSensorId.513106929 = nbAlinkEnc_0_3_SPLK
otherStateSensorId.2637721517 = nbAlinkEnc_0_4_SMOK
otherStateSensorId.3502248167 = nbLinkStatus_eth0
otherStateSensorId.3823829717 = nbAlinkPowerStatus_can0
otherStateSensorValue.513106929 = 0
otherStateSensorValue.2637721517 = 0
otherStateSensorValue.3502248167 = 1
otherStateSensorValue.3823829717 = 0
otherStateSensorErrorStatus.513106929 = 0
otherStateSensorErrorStatus.2637721517 = 0
otherStateSensorErrorStatus.3502248167 = 0
otherStateSensorErrorStatus.3823829717 = 0
otherStateSensorLabel.513106929 = Spot leak  (3)
otherStateSensorLabel.2637721517 = Smoke  (4)
otherStateSensorLabel.3502248167 = Ethernet Link Status
otherStateSensorLabel.3823829717 = A-Link Bus Power
otherStateSensorEncId.513106929 = nbAlinkEnc_0
otherStateSensorEncId.2637721517 = nbAlinkEnc_0
otherStateSensorEncId.3502248167 = nbBaseEnclosure
otherStateSensorEncId.3823829717 = nbBaseEnclosure
otherStateSensorPortId.513106929 =
otherStateSensorPortId.2637721517 =
otherStateSensorPortId.3502248167 =
otherStateSensorPortId.3823829717 =
otherStateSensorValueStr.513106929 = No Leak
otherStateSensorValueStr.2637721517 = No Smoke
otherStateSensorValueStr.3502248167 = Up
otherStateSensorValueStr.3823829717 = OK
otherStateSensorEntry.8.513106929 = 513106929
otherStateSensorEntry.8.2637721517 = 2637721517
otherStateSensorEntry.8.3502248167 = 3502248167
otherStateSensorEntry.8.3823829717 = 3823829717
```

File: doc/Developing/Sensor-State-Support.md [hunk 2]
```diff
@@ -48,6 +48,66 @@ We also map these values to the actual state sensor(state_index) where these val
...
+                descr: cameraMotionSensorLabel
+                index: 'cameraMotionSensor.{{ $index }}'
+                state_name: cameraMotionSensor
+                states:
+                    - { descr: 'null', graph: 0, value: -1, generic: 3 }
+                    - { descr: noMotion, graph: 0, value: 0, generic: 0 }
+                    - { descr: motionDetected, graph: 0, value: 1, generic: 2 }
+            -
+                oid: otherStateSensorTable
+                value: otherStateSensorValue
```
Comment (Rosiak) [inline doc/Developing/Sensor-State-Support.md]: I believe the correct name is:
otherStateSensorErrorStatus

File: doc/Developing/Sensor-State-Support.md [hunk 3]
```diff
@@ -48,6 +48,66 @@ We also map these values to the actual state sensor(state_index) where these val
...
+                num_oid: .1.3.6.1.4.1.5528.100.4.2.10.1.3.
+                descr: otherStateSensorLabel
+                index: '{{ $index }}'
+                state_name: otherStateSensorErrorStatus
```
Comment (Rosiak) [inline doc/Developing/Sensor-State-Support.md]: How about otherStateSensor?

File: includes/definitions/discovery/netbotz.yaml [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+                descr: cameraMotionSensorLabel
+                index: 'cameraMotionSensor.{{ $index }}'
+                state_name: cameraMotionSensor
+                states:
+                    - { descr: 'null', graph: 0, value: -1, generic: 3 }
+                    - { descr: noMotion, graph: 0, value: 0, generic: 0 }
+                    - { descr: motionDetected, graph: 0, value: 1, generic: 2 }
+            -
+                oid: otherStateSensorTable
+                value: otherStateSensorValue
```
Comment (Rosiak) [inline includes/definitions/discovery/netbotz.yaml]: I believe the correct name is:
otherStateSensorErrorStatus

File: includes/definitions/discovery/netbotz.yaml [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+                num_oid: .1.3.6.1.4.1.5528.100.4.2.10.1.3.
+                descr: otherStateSensorLabel
+                index: '{{ $index }}'
+                state_name: otherStateSensorErrorStatus
```
Comment (Rosiak) [inline includes/definitions/discovery/netbotz.yaml]: How about otherStateSensor?

File: doc/Developing/Sensor-State-Support.md (hunk 3 above)
Comment (laf) [inline doc/Developing/Sensor-State-Support.md]: It's how it is in the old code so replicating that.

File: doc/Developing/Sensor-State-Support.md (hunk 2 above)
Comment (laf) [inline doc/Developing/Sensor-State-Support.md]: Updated, don't think it matters though.

**Pull request merged**